### ISSUE TITLE
[BUG] MTP `--enable-mtp` silently no-op on MLLM path (Qwen3.6-27B-bf16): speed unchanged from baseline

### ISSUE THREAD

**Issue by hailanlan0577** [waybarrios/vllm-mlx#471]
## Context                                                                                                                                               
                                                                                                                                                           
  Testing MTP speculative decoding on a Qwen3.6-27B-bf16 VLM (Dense, not MoE) running through the `--mllm` path on Apple Silicon. MTP injection succeeds   
  end-to-end and the server logs `MTP: enabled, draft_tokens=5`, but **wall-clock decode speed is identical with or without `--enable-mtp`** — strongly    
  suggesting the MLLM MTP step is not actually plugged into the generator loop.                                                                            
                                                                                                                 
  ## Environment

  - **Hardware**: Mac Studio M3 Ultra, 256 GB unified memory                                                                                               
  - **vllm-mlx**: master @ `33969810` (commit msg: "Fix/test stream thread safety additions (#467)", dated 2026-04-29; installed via tarball, since PyPI
  0.2.9 lacks this fix)                                                                                                                                    
  - **mlx-vlm**: 0.4.4, **mlx-lm**: 0.31.3, **mlx**: 0.31.2                                                      
  - **Model**: `mlx-community/Qwen3.6-27B-bf16` (Dense VL, `model_type: qwen3_5`, `hidden_size: 5120`, `num_hidden_layers: 64`, `mtp_num_hidden_layers: 1`)
  - **Source for MTP shards**: `Qwen/Qwen3.6-27B` (bf16 official), 15 mtp.* keys in shards 13 & 15                                                         
  - **MTP injection**: ran `scripts/add_mtp_weights_qwen35.py` (with default 4-bit quantization), produced 314 MB `mtp/weights.safetensors` (29 quantized  
  tensors). `config.json` updated with `num_nextn_predict_layers: 1`.                                                                                      
                                                                                                                                                           
  ## Server start command                                                                                                                                  
                                                                                                                                                           
  ```bash
  vllm-mlx serve /path/to/Qwen3.6-27B-bf16 \                                                                                                               
    --port 8081 --host 0.0.0.0 \                                                                                                                           
    --mllm --trust-remote-code \
    --disable-prefix-cache \                                                                                                                               
    --continuous-batching --use-paged-cache --max-num-seqs 10 \                                                                                          
    --enable-mtp --mtp-num-draft-tokens 5                                                                                                                  
  ```                                                                                                                                                      
                                                                                                                                                           
  ## Startup logs (looks healthy)                                                                                                                          
                                                                                                                                                         
  ```
  INFO:vllm_mlx.patches.qwen3_5_mtp:[MTP inject] Building TextModelArgs from config dict
  INFO:vllm_mlx.patches.qwen3_5_mtp:[MTP inject] Creating MTP module (1 layers, Dense)                                                                     
  INFO:vllm_mlx.patches.qwen3_5_mtp:[MTP inject] Loading weights from weights.safetensors (BF16, no quantization)                                          
  INFO:vllm_mlx.patches.qwen3_5_mtp:[MTP inject] Loaded 15 MTP weight tensors (dequantized from quantized)                                                 
  INFO:vllm_mlx.patches.qwen3_5_mtp:[MTP inject] Model class patched with MTP support                                                                      
  INFO:vllm_mlx.engine.batched:[MTP-MLLM] Qwen3.5 MTP injected successfully                                                                                
  WARNING:vllm_mlx.mllm_batch_generator:[MTP-MLLM] num_draft_tokens=5 requested,                                                                           
    but the current batched MLLM MTP path drafts exactly one token per verify step                                                                         
  INFO:vllm_mlx.mllm_batch_generator:[MTP-MLLM] installed with                                                                                             
    num_draft_tokens=5, effective_draft_tokens=1, always-advance verified mode                                                                             
  MTP: enabled, draft_tokens=5                                                                                                                             
  ```                                                                                                                                                      
                                                                                                                                                           
  The `[Qwen3.5 patch] Attention patched for BatchKV Cache support` line is also logged 64× (one per layer), so attention monkey-patching ran.
                                                                                                                                                           
  ## Benchmark (3 single-stream runs, max_tokens=250, temperature=0)
                                                                                                                                                           
  | Configuration | Run 1 | Run 2 | Run 3 | Avg |                                                                
  |---|---|---|---|---|                                                                                                                                    
  | **Baseline (no `--enable-mtp`)** | 11.94 t/s | 12.69 t/s | 12.65 t/s | **12.43 t/s** |                       
  | **With `--enable-mtp --mtp-num-draft-tokens 5`** | 11.94 t/s | 12.69 t/s | 12.65 t/s | **12.43 t/s** |

  **Wall-clock identical to the millisecond.** Even given `effective_draft_tokens=1` (down from 5) and `always-advance` (no verify), one would expect at
  least *some* speedup from skipping the verify pass.

  ## Hypothesis

  Reading `vllm_mlx/scheduler.py` (`_install_mtp`, `_mtp_step`) vs `vllm_mlx/mllm_scheduler.py` and `vllm_mlx/mllm_batch_generator.py`, the LLM path looks
  fully wired but the MLLM path appears to install MTP attention monkey-patches without actually invoking `mtp_forward` from the generator step. The
  "always-advance verified mode" warning suggests this is a known placeholder.

  ## Questions

  1. Is MLLM-path MTP intentionally a no-op placeholder pending future work, or should it produce *any* speedup today?
  2. Is there a tracking issue for completing MLLM MTP wiring? Happy to test patches.
  3. Would `--no-quantize` MTP weights change behavior here, or is the issue strictly in the generator step plumbing (independent of weight precision)?

  Happy to provide more logs, run additional configurations, or test patches against this exact setup.

**Comment by Thump604** [comment]
Thanks for the detailed repro. I agree this should not present as a working speedup just because MTP weights inject successfully.

Current read: on the MLLM path, startup can report the requested draft token count, but the runtime path is effectively constrained to one draft per verify step and is bypassed for multi-request batches, non-greedy sampling, and active logits processors. That means `--enable-mtp --mtp-num-draft-tokens 5` can look enabled while producing no meaningful wall-clock improvement.

I opened #473 as the first fix. It does not claim to make MLLM MTP faster. It makes the behavior observable and honest by exposing requested vs effective draft count plus attempts/accepted/rejected/errors in `/v1/status`, and by changing startup logging so requested draft tokens are not confused with effective speculative work.

After that lands, the next step is an on/off MLLM MTP performance pass using those counters. I do not think `--no-quantize` is the first variable to test here; the main issue is generator-path behavior and observability, not weight quantization.

**Comment by Thump604** [comment]
One additional data point from review on #473: if the MTP weights were extracted with the default quantization path (e.g., `scripts/add_mtp_weights_qwen35.py` without `--no-quantize`), the draft weights are 4-bit quantized. Quantization error in the draft logits makes every draft token wrong, resulting in 0% acceptance rate regardless of serving configuration.

The fix is to extract BF16 MTP weights from the original HF model (e.g., `Qwen/Qwen3.6-27B`) using `--no-quantize`. With BF16 MTP weights, acceptance rates of 78-85% are expected on Qwen 3.5/3.6 models. See https://github.com/vllm-project/vllm/issues/36331 for background.

This is separate from the MoE-to-dense MLLM path bug (mlx-vlm#1073), which also makes MTP ineffective on the MLLM path. Both issues need to be resolved for MTP to work correctly on Qwen MoE models served through mlx_vlm.

**Comment by Thump604** [comment]
One important test-control issue: the repro path here appears to use 4-bit quantized MTP weights from scripts/add_mtp_weights_qwen35.py default behavior.

That can make MTP look broken even when the runtime path is doing what it was asked to do, because quantized draft weights can drive acceptance to 0%. For a clean MTP performance/acceptance test on Qwen3.5/Qwen3.6, the draft weights should be extracted in native BF16 with --no-quantize from the original HF model, then the run should compare:

- baseline without MTP
- MTP with quantized draft weights, expected to show poor or zero acceptance
- MTP with BF16 draft weights, expected to show meaningful acceptance if the rest of the path is correct

PR #473 does not fix MTP performance. It makes the MLLM path report requested vs effective draft tokens, attempts, acceptance, errors, and bypass counters so this kind of test can be interpreted without reading the source.

**Comment by Thump604** [comment]
Current source triage against `main` shows that the original "MLLM MTP never invokes the generator path" diagnosis is no longer reproducible from the implementation.

- `vllm_mlx/mllm_batch_generator.py:install_mtp_mllm` installs `_mtp_step`, and that step calls `language_model.mtp_forward(...)` before target verification.
- `tests/test_mllm_continuous_batching.py::test_install_mtp_mllm_accepted_drafts_bypass_request_sampler` exercises that path and asserts `attempted == 1`, `accepted == 1`, and a 1.0 acceptance rate.
- #473, #504, and #527 are merged. `/v1/status.mtp` now distinguishes attempted work from bypasses and reports requested versus effective draft count.

The current limitation is explicit rather than silent: the merged MLLM path drafts one token per verification step and bypasses MTP for prefill, concurrent batches, non-greedy sampling, and active request-local logits processors. A sampled/concurrent MTP verifier is separate feature work and needs its own current benchmark and correctness evidence.

I am not claiming a present performance qualification from this source review. The April benchmark used an older environment and cannot establish current throughput. Closing this issue because its stated no-op code-path diagnosis is superseded; a new issue should include a current reproducible workload and `/v1/status.mtp` counters if the present guarded path is not engaging as expected.
